Cumulus 16 support

colecu opened the pull request:
NSIDC has a need to work with cumulus-api in a Cumulus 16 environment. Between Cumulus 15 and 16 there are some breaking changes involving existing `granules` PUT endpoints. PUT was deprecated in favor of PATCH for existing functionality, with plans for PUT behavior to transition to fully overwriting the target. There was also a requirement added 
to provide a version header with these requests.

These changes have been implemented in an NSIDC fork of the repository, and we are providing the corresponding PR should it be useful for any updates you have planned for the next PI.

colecu committed:
Fix package dependency issue

It appears that an incompatibility was introduced between the
`cryptography` and `pyOpenSSL` packages. Running the code with the
previous versions resulted in an error reporting that module `lib`
did not have an attribute named `OpenSSL_add_all_algorithms`. This
was addressed by moving to the latest version of `pyOpenSSL`.

More information:
https://levelup.gitconnected.com/fix-attributeerror-module-lib-has-no-attribute-openssl-521a35d83769

colecu committed:
Update for Cumulus 16.x

As of Cumulus 16, a `Cumulus-API-Version` header is required for
all PUT and PATCH requests in order to avoid unexpected destructive
writes due to changes in PUT handling. Further PUT was deprecated
in favor of PATCH for previous behavior in version 14.1.0 with
advice to being changing existing PUT requests to PATCH requests.
As of version 16, continued use of PUT could result in unexpected
behavior. See Cumulus changelog and CUMULUS-3072 for additional
information.

This commit adds the `Cumulus-API-Version` header with a value of
`2` to all requests and replaces all `put` verbs with `patch`
verbs.

camposeddie commented on cumulus_api/main.py:
```diff
@@ -33,7 +33,10 @@ def __init__(self, use_os_env: bool = True, config_path: str = None, token: str
-        self.HEADERS = {'Authorization': f'Bearer {self.TOKEN}'}
+        self.HEADERS = {
+            'Authorization': f'Bearer {self.TOKEN}',
+            'Cumulus-API-Version': '2',
```
This header value should probably be implemented at the individual method level since most don't currently require it and there is potential for this to be a different value for different methods in the future (assumption given that some endpoints are currently at "1" while others are at 2.

colecu replied:
Based on the current Cumulus 16 implementation, I would have to agree. I took a closer look at the middleware code that is being used for version checking, and it is written such that each endpoint could provide an arbitrary version number. I don't know if there is any intent to use it this way, but it is supported.

At a first look, it appears that some refactoring would be required in order to apply the version header on an individual method level, so I am leaving that at the instance level until we get some more feedback from Cumulus.

camposeddie replied:
Discussion with Jonathan from Cumulus team: https://eosdis.slack.com/archives/C76M787MZ/p1691176763321379?thread_ts=1691084537.427589&cid=C76M787MZ 
There shouldn't be an issue with including this header for endpoints which don't need them, so we will adapt this approach to avoid unnecessary updates on a per method basis.

camposeddie commented on cumulus_api/main.py:
```diff
@@ -131,7 +134,7 @@ def update_provider(self, data):
         """
         record_type = f"providers/{data['id']}"
-        return self.__crud_records(record_type=record_type, verb="put", data=data)
+        return self.__crud_records(record_type=record_type, verb="patch", data=data)
```
update provider still uses PUT: https://github.com/nasa/cumulus/blob/v16.0.0/packages/integration-tests/api/api.js#L334

colecu replied:
The wording in the Cumulus change log, as well as a Slack thread that I pulled up sounded very definitive that *all* uses of PUT should be transitioned to PATCH. However you are correct that Cumulus hasn't done that in v16 api-client. My interpretation was that they added identical support for both PUT and PATCH to allow for the transition until the PUT implementations actually start diverging. But that doesn't appear to be the case. I will update our branch and PR appropriately.

colecu replied:
Reverted.

camposeddie commented on cumulus_api/main.py:
```diff
@@ -190,7 +193,7 @@ def update_collection(self, data):
         """
         record_type = f"collections/{data['name']}/{data['version']}"
-        return self.__crud_records(record_type=record_type, verb="put", data=data)
+        return self.__crud_records(record_type=record_type, verb="patch", data=data)
```
update collection still uses PUT: https://github.com/nasa/cumulus/blob/v16.0.0/packages/api-client/src/collections.ts#L63

colecu replied:
Reverted.

camposeddie commented on cumulus_api/main.py:
```diff
@@ -406,7 +409,7 @@ def update_rule(self, data):
         """
         record_type = f"rules/{data['name']}"
-        return self.__crud_records(record_type=record_type, verb="put", data=data)
+        return self.__crud_records(record_type=record_type, verb="patch", data=data)
```
update rule still uses PUT: https://github.com/nasa/cumulus/blob/v16.0.0/packages/api-client/src/rules.ts#L66

colecu replied:
Reverted

camposeddie commented on cumulus_api/main.py:
```diff
@@ -425,7 +428,7 @@ def run_rule(self, rule_name):
         record_type = f"rules/{rule_name}"
         data = {"name": rule_name, "action": "rerun"}
-        return self.__crud_records(record_type=record_type, verb="put", data=data)
+        return self.__crud_records(record_type=record_type, verb="patch", data=data)
```
rerun rule still uses PUT: https://github.com/nasa/cumulus/blob/v16.0.0/packages/api-client/src/rules.ts#L66

colecu replied:
Reverted.

camposeddie commented on cumulus_api/main.py:
```diff
@@ -539,7 +542,7 @@ def update_execution(self, data):
         """
         record_type = f"executions/{data['arn']}"
-        return self.__crud_records(record_type=record_type, verb="put", data=data)
+        return self.__crud_records(record_type=record_type, verb="patch", data=data)
```
update execution still uses PUT: https://github.com/nasa/cumulus/blob/v16.0.0/packages/api-client/src/executions.ts#L147

colecu replied:
Reverted.

camposeddie commented on cumulus_api/main.py:
```diff
@@ -683,7 +686,7 @@ def update_elasticsearch_index(self, data):
         """
         record_type = "elasticsearch/change-index"
-        return self.__crud_records(record_type=record_type, verb="put", data=data)
+        return self.__crud_records(record_type=record_type, verb="patch", data=data)
```
change index still uses PUT: https://nasa.github.io/cumulus-api/unreleased/#change-index

colecu replied:
Reverted

camposeddie commented:
- The granule endpoint updates look good 👍 
- The other "update" endpoints weren't actually updated to use PATCH and still rely on PUT in v16.
- The api version header value should probably be method-based. Currently v2 is the only api version option; however, my understanding is that this implies the endpoints not using the header are v1 which arises the concern that the cumulus api can have multiple endpoint api versions for a single Cumulus version. Given this, more discussion is needed with the Cumulus team on understanding this approach and what would be considered best practice here.
- In general, we should tag up with Cumulus team about migrating "update" endpoints to actually update with PATCH verbiage and creating "replace" endpoints which actually replace with PUT verbiage.

I'm going to leave this PR open until further discussion with the Cumulus team has completed.

colecu committed:
Revert "Update for Cumulus 16.x"

Only the `granules` endpoint should have been updated from PUT to
PATCH. The previous commit was prompted by a misunderstanding of a
Cumulus CHANGELOG entry and Slack thread.

This reverts commit 458ef8575122bfb7b0488cae25e825c8f440b79d.

colecu committed:
Update for Cumulus 16.x

As of Cumulus 16, a `Cumulus-API-Version` header is required for
all `granules` PUT and PATCH requests in order to avoid unexpected
destructive writes due to changes in PUT handling. Further PUT was
deprecated in favor of PATCH for `granules` requests in version
14.1.0 with advice to begin changing existing PUT requests to PATCH
requests. As of version 16, continued use of PUT for `granules`
could result in unexpected behavior. See Cumulus changelog and
CUMULUS-3072 for additional information.

This commit adds the `Cumulus-API-Version` header with a value of
`2` to all requests and replaces all `put` verbs for `granules`
requests with `patch` verbs.

lindsleycj commented:
@camposeddie Hi, wondering what the status of this PR is?  We have a need to update granules with Cumulus v16 api.

colecu committed:
Merge remote-tracking branch 'origin/master' into cumulus-16-support

colecu commented:
Updates in master have been merged to the PR branch.

Michaellh0079 reviewed (approved):
Tested with pylot, seems to work as expected.
